### PR TITLE
Fixed #28342 -- Fixed PyLibMCCache error handling

### PR DESCRIPTION
**Pull request by julianandrews**
https://code.djangoproject.com/ticket/28342

### CODE REVIEW

**Comment by timgraham** [comment]
@edmorley, you've made some contributions for pylibmc, do you have any feedback about this?

More tests are needed to complete the test coverage, but let's make sure the idea is okay first.

**Comment by edmorley** [comment]
Hi!

My initial thoughts:
* I'm a fan of upstreaming more of django-pylibmc to Django (django-pylibmc/django-pylibmc#37) - ideally such that it's not required at all in the future. (cc @jwhitlock)
* It definitely makes sense for the user-facing behaviour to be the same for both the python-memcached and pylibmc backends
* I suspect moving the python-memcached backend to raising would cause more breakage so probably not viable (for a start, new generic exceptions would be required to replace pylibmc's ones, so even existing pylibmc users would be broken too) - so the direction this PR is taking makes most sense.

**Comment by jwhitlock** [comment]
I haven't been an active maintainer or user of django-pylibmc for a few years, so my memory is hazy.  I recall having trouble with a Heroku app, which uses config by environment variables.  I had an error in my config string parsing, and connections were failing.  Because the the exceptions were caught, my cache calls were silently failing. An "add" silently did nothing, "get" returned a cache miss, and my app was slower than expected.  Once I figured out what was going on, it was tricky to setup logging so that I could see the error messages with Heroku tools.  I would have preferred raw exceptions, and was thinking of making this a configuration option for django-pylibmc.

I haven't encountered issues like @julianandrews, but I'd expect some users like me want noisy failures, and others want silent or logging failures.  This desire may change based on the code being in development or production.

**Comment by julianandrews** [comment]
@jwhitlock It's a good point about the logging and the exceptions. It occurs to me that this might be a natural use case for the `warnings` module. If a cache failure raised a warning, then a developer could use warning filters to turn the cache warnings into exceptions for development, or the `logging.captureWarnings` method to redirect the warnings to their logging pipeline.

As a plus, I think that this would be reasonably backwards compatible. Some sites would start to see warnings, and sites already turning warnings into exceptions might see some unexpected failures, but I can't imagine that's a very common practice in production.

@timgraham I'd be happy to make that change if you think that's a good idea (along with a backwards-incompatible changelog entry).

Also, I'd be happy to add more tests. I wasn't sure how many tests were appropriate. I thought about adding tests for server failure for each method, but I hesitated because such a large fraction of the `pylibmc` backend tests would have ended up testing my little edge case. Guidance on this front would be welcome!

**Comment by timgraham** [comment]
I don't see any obvious problems with the approach you suggested. As for tests, the idea is to have 100% coverage to that changes aren't inadvertently removed in the future. Perhaps using a loop and `subTest` would allow making the tests less repetitive.

File: django/core/cache/backends/memcached.py [hunk 1]
```diff
@@ -194,6 +194,56 @@ def __init__(self, server, params):
     def _cache(self):
         return self._lib.Client(self._servers, **self._options)
 
+    def add(self, key, value, timeout=DEFAULT_TIMEOUT, version=None):
```
Comment (edmorley) [inline django/core/cache/backends/memcached.py]: Since the args/kwargs aren't used directly, would this be simpler to maintain as `*args, **kwargs` instead? (I can just see a case where it ends up out of sync with `BaseMemcachedCache`)

File: django/core/cache/backends/memcached.py [hunk 2]
```diff
@@ -194,6 +194,56 @@ def __init__(self, server, params):
...
+        try:
+            return super().add(key, value, timeout=timeout, version=version)
+        except (self._lib.ConnectionError, self._lib.ServerDown, self._lib.Error) as e:
+            warnings.warn(str(e), RuntimeWarning)
+            return value
+
+    def get(self, key, default=None, version=None):
```
Comment (edmorley) [inline django/core/cache/backends/memcached.py]: `get_many` and `clear` aren't subclassed here, should they be?

Also, I was wondering if it might be easier to keep in sync in the future, if the order of methods matched those from `BaseMemcachedCache`?

File: tests/cache/tests.py [hunk 3]
```diff
@@ -1312,6 +1312,39 @@ def test_pylibmc_legacy_options(self):
...
 
+    # We want to test what happens if memcached is down. 192.0.2.0 is reserved
+    # for examples, and therefore should never point to a valid memcached
+    # server. See https://tools.ietf.org/html/rfc5737.
+    @override_settings(CACHES=caches_setting_for_tests(
+        base=PyLibMCCache_params,
+        exclude=memcached_excluded_caches,
+        LOCATION='192.0.2.0',
+    ))
+    def test_pylibmc_warnings(self):
```
Comment (edmorley) [inline tests/cache/tests.py]: A few of the methods aren't tested here - perhaps Tim's loop idea above might make it easier to test them all without the duplication?

File: django/core/cache/backends/memcached.py (hunk 1 above)
Comment (edmorley) [reply django/core/cache/backends/memcached.py]: Oh `value` is used, I can't read! (There are some methods that don't use the passed arguments, but perhaps they should all be specified for consistency)

**Comment by timgraham** [comment]
Closing due to inactivity.
